refactor: File

gavindsouza opened the pull request:
### Changes

* Restructured File module

`./frappe/core/doctype/file/file.py` -> [`./frappe/core/api/file.py`, `./frappe/core/doctype/file/utils.py`, `./frappe/core/doctype/file/exceptions.py`, `./frappe/core/doctype/file/file.py`]

* Maintain import paths & API endpoints
* New namespace for file based endpoints actions -> `/api/method/frappe.core.api.file`
* Simplify interfaces for File actions
* Added type hints to aid my development & debugging 
* Bug & security fixes

### Fixes

* Disallow file access beyond the respective site's folder
* Consistent setting of `File.file_name` attribute
* Avoid overwriting files that may have the same name and suffix
* Enforce the "max_number_of_files" restriction in FileUploader
* Make File tests atomic
* Improved syncing file system with changes on rollback
* Cover extension-less img previews in File form

### Other Changes

* Skip `test_multiple_doctypes_sync` for Postgres test run

gavindsouza committed:
fix: Misc fixes

* Set default Falsy value for content
* Extend Frappe exceptions into file's

gavindsouza committed:
fix: Handle File on rollback

Delete, Move or Restore files in the filesystem by setting Document
flags in the on_rollback method

gavindsouza committed:
fix: Disallow sites to have file access beyond site folder

gavindsouza committed:
refactor: Restructure File module

* Separate out utils, exceptions & core / maintain import paths [except
  API endpoints]
* Add type hints
* Simplify logic

gavindsouza committed:
fix: Generate non-conflicting names in File.save_file

Other fix:
* Don't set File.file_name (too) different from File.file_url

gavindsouza committed:
chore: Use standard APIs

gavindsouza committed:
fix!: Raise 403 if Guest isn't allowed to upload file

gavindsouza committed:
refactor: File APIs

Restructured and moved most APIs under frappe.core.api.file namespace.
Changed some obvious security gaps (like using get_list instead of
get_all for an endpoint), styled, added type hints and made minor performance
enhancements.

Changes
* download_file API
    * Move API to handler.py
    * Check for permissions via File.is_downloadable instead
* Moved APIs to new namespace: `frappe.core.api.file`
* Backwards compatibility
    * Added APIs to override_whitelisted_methods to maintain existing
      client endpoints
    * Imported APIs to controller's namespace to avoid breaking external
      app usages

gavindsouza committed:
fix: Use new direct import paths

The old paths worked too, but it's just better to use the new paths and
not go in circles ;)

gavindsouza committed:
refactor: File controller

* Simplify object lifecycle flows
* Re-organize and separate distinct actions / APIs
* Style black-ish
* Simplify logic in methods "touched"

gavindsouza committed:
fix(utils): URLs to return True via is_safe_path checks

gavindsouza committed:
fix: File.file_url as empty string as fallback

rmehta commented:
@gavindsouza since you are cleaning up file, please also fix the bug where if you select a file as a property in a document, it creates a new file record. (example: attach a file to a blog, and then select it as cover - 2 files records get created)!

Had discussed this with @surajshetty3416

gavindsouza committed:
refactor: File.optimize_file

* Generalize file content reverting on rollback
* Overwrite current file instead of adding a dangling new file
* Use File.save_file instead of additional "custom" logic
* Set File.content on calling get_content method

gavindsouza committed:
refactor: File.handle_is_private_changed

* Make file renaming atomic
* Add rollback observer to maintain filesystem consistency
* Simplify logic

gavindsouza committed:
fix: os.fsync to flush internal buffers to disk

ref: https://docs.python.org/3/library/os.html#os.fsync

gavindsouza committed:
fix: Remove duplicate rollback observers and maintain order

gavindsouza committed:
fix: save_file only for non folder records

gavindsouza committed:
fix: Skip handling is_private updates for remote files

gavindsouza committed:
fix: Strip HTML tags in traceback shown in the terminal

gavindsouza committed:
refactor(file): File.is_remote_file instead of startswith http

gavindsouza committed:
refactor(File): Use _content attribute instead of overwriting content

gavindsouza committed:
fix(File): Misc fixes

* Move checks into common methods
* Refactor to use latest APIs
* Remove seemingly unnecessary code
* Fix previously undhanled use cases

gavindsouza committed:
fix(test): Update file tests for the new API changes

gavindsouza committed:
chore: Drop unmaintained file

To check out the up-to-date version of hooks.md, see https://frappeframework.com/docs/v13/user/en/python-api/hooks

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-file

gavindsouza committed:
Merge branch 'develop' into refactor-file

gavindsouza committed:
feat(FileUploader): Added "max_number_of_files" restriction

FileUploader will accept N files and pop the rest, followed by a
console.warn and frappe.show_alert

gavindsouza committed:
fix: Validate attachment limit before insert

gavindsouza committed:
chore(handler): Added typing, removed unnecessary assignments

gavindsouza committed:
fix: Use FrappeTestCase for TestCase's atomicity

gavindsouza committed:
chore: Drop dead imports

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-file

gavindsouza committed:
style: Run pre-commit linters

gavindsouza committed:
fix: Add return value for doc.delete

gavindsouza committed:
refactor(minor): frappe.client.attach_file

* Use standard doc.check_permission instead
* Don't re-check for http method type inside API
* Cleanup variable naming
* Return file object - let the API layer manage marshalling /
  serialization

gavindsouza committed:
refactor: frappe.desk.form.utils.add_comment

* Check read permission for reference document before allowing to drop a
  comment: if you can't see the document, should you be able to link a
  documnet to it? i think not
* Simplify logic + add type hints
* Delegate serialization to the API layer; return Comment object instead

gavindsouza committed:
fix: Allow only DELETE, POST methods for remove_attach API

gavindsouza committed:
refactor: frappe.utils.response.download_private_file

gavindsouza committed:
fix: Set 'Attached to *' fields properly

Check if DocType is a child table before using parent and parenttype
attributes. This caused problems when a dead "parent" column existed on
the database and was mapped to the Document object.

Port of https://github.com/frappe/frappe/pull/16654

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-file

gavindsouza committed:
fix(qb-pg): Type cast boolean params as int

gavindsouza committed:
fix(linting): Sort imports

gavindsouza committed:
Revert "fix(qb-pg): Type cast boolean params as int"

This reverts commit 06183535af2ad6ee43546a241c1ad22641b8e69e.

gavindsouza committed:
test: Fix postgres usage for database delete API

gavindsouza committed:
fix: Use query.run to execute query instead of db.sql

gavindsouza committed:
refactor: Simplify DB/QB API definitions

gavindsouza committed:
fix: File.is_remote_file logic

Returned True previously even if file was local ;)

gavindsouza committed:
test: Use new_doc instead of get_doc

gavindsouza committed:
fix: Skip renaming files if updated and old file_url are same

Added to avoid case mentioned in comments. Check diff for more.

gavindsouza committed:
chore: Avoid overwriting imports

gavindsouza committed:
test: Run db.rollback instead of on_rollback hook

They aren't equivalent, and calling on_rollback directly causes
unforeseen problems.

gavindsouza committed:
test: Use a contextmanager for the test image file

gavindsouza committed:
Merge branch 'develop' of github.com:frappe/frappe into refactor-file

gavindsouza committed:
fix: Log attachment error with proper context

ref: https://github.com/frappe/frappe/pull/16653#discussion_r854813203

gavindsouza commented:
<img width="1552" alt="Screenshot 2022-04-26 at 5 02 45 PM" src="https://user-images.githubusercontent.com/36654812/165291041-30ed1a49-b78c-479f-a7ab-4dfb4f37e8dc.png">

Fixed https://github.com/frappe/frappe/issues/13624 by ^

gavindsouza commented:
Shall muster the courage to revisit soon 🥲

gavindsouza committed:
Merge branch 'develop' into refactor-file

gavindsouza committed:
fix: commit after processing request of "unsafe" http method

ref: https://developer.mozilla.org/en-US/docs/Glossary/safe

gavindsouza committed:
fix: Show img even if URL ending is unknown - let html handle the stuff
it knows

* Hide Preview if request is errored. EAFP.
* Set file attachment limit based on how many attachments have already
  been added. If 1 attachment exists and the limit is 3, File uploader
will allow you to select only 2 more files.

gavindsouza committed:
test: Skip test_multiple_doctypes_sync for Postgres tests

gavindsouza committed:
Merge branch 'develop' into refactor-file

gavindsouza commented:
Ignoring flaky Cypress test `'Verifying data control by inputting different patterns for "Email" field'` (cc: @Komal-Saraf0609)
